pib/remote.py

```diff
@@ -0,0 +1,124 @@
+import boto3
+import botocore.exceptions
+import json
+import os
+import shlex
+from pathlib import Path
+from subprocess import Popen, PIPE, CalledProcessError
+
+
+def get_aws_acct_id():
+    return boto3.client('sts').get_caller_identity().get('Account')
+
+
+def get_env():
+    branch = os.getenv('TRAVIS_BRANCH', '').lower()
+    if branch.startswith('env/'):
+        return branch.split('/')[1]
+    else:
+        raise ValueError('Not an environment branch. Expected format env/$name but found {0}'.format(branch))
+
+
+def merge_dicts(defaults, overrides):
+    res = {}
+    res.update(defaults)
+    res.update(overrides)
+    return res
+
+
+def run(cmd, allowed_exit_codes={0}, **kwargs):
+    if type(cmd) is str:
+        cmd = shlex.split(cmd)
+
+    merged = merge_dicts({'stdout': PIPE, 'bufsize': 1, 'universal_newlines': True}, {})
+    with Popen(cmd, **merged) as proc:
+        for line in proc.stdout:
+            print(line, end='')
+
+    status = proc.returncode
+    if status not in allowed_exit_codes:
+        raise CalledProcessError(status, cmd, None)
+
+    return proc.returncode
+
+
+class RemoteDeploy:
+
+    def __init__(self, config):
+        self.config = config
+        self.prepared = False
+
+    def prepare(self):
+        """Perform preparations before performing any type of remote operations, for example, fetching remote Terraform
+        state or downloading required modules."""
+        state_bucket = self.config.get('state_bucket')
```

Comment:
`get()` will return `None` if key is missing... might be better to do `config['state_bucket']` so it errors out early if required value is missing.